torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
+def all_eq(n: Node):
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        args1 = arg_type.__args__
+        args2 = n.type.__args__
+        return [Equality(args1[i], args2[i]) for i in range(len(args1))]
+
+
+@register_refinement_rule(torch.nn.MaxPool2d)
+def first_two(n: Node):
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        args1 = arg_type.__args__
+        args2 = n.type.__args__
+        return [Equality(args1[0], args2[0]), Equality(args1[1], args2[1])]
+
+@register_refinement_rule(torch.flatten)
+def flatten_refinement_rule(n: Node):
+    assert isinstance(n.args[0], Node)
+
+    eq_const = []
+
+    start_dim = 1
+    end_dim = -1
+
+    if len(n.args) > 1:
+        assert isinstance(n.args[1], int)
+        start_dim = n.args[1]
+
+    if len(n.args) > 2:
+        assert isinstance(n.args[2], int)
+        end_dim = n.args[2]
+
+    if isinstance(n.type, TensorType) and isinstance(n.args[0].type, TensorType):
+        l = len(n.type.__args__)
+        start_dim = l if start_dim == -1 else start_dim
+        end_dim = l + end_dim + 1 if end_dim < 0 else end_dim + 1
+
+        for t1, t2 in zip(n.type.__args__[0:start_dim], n.args[0].type.__args__[0:start_dim]):
+            eq_const.append(Equality(t1, t2))
+
+        for t1, t2 in zip(n.type.__args__[start_dim:end_dim], n.args[0].type.__args__[start_dim:end_dim]):
+            eq_const.append(Equality(t1, t2))
+
+        return eq_const
+
+class Refine:
+    """
+    Symbolic shape inference.
+    Generates constraints over type variables.
+    Currently all constraints are equality constraints.
+    """
+    def __init__(self, traced):
+        self.constraints = []
+        self.traced = traced
+        self.curr_symbol = self.symbol_gen()
+
+    def symbol_gen(self):
+        val = [0]
+
+        def inc():
+            val[0] += 1
+            return val[0]
+        return inc
+
+    def refine(self):
+        """
+        Generates constraints for
+        every node in the graph based on
+        the operation.
+        """
+        graph = self.traced.graph
+        for n in graph.nodes:
+            self.refine_node(n)
+        return True
+
+    def replace_dyn_with_fresh_var(self, typ):
+        """
+        Replace all unknown types with fresh type variables.
+        """
+        if typ == Dyn:
+            new_symbol = Var(self.curr_symbol())
+            return new_symbol
+        elif isinstance(typ, TensorType):
+            new_args = []
+            for a in typ.__args__:
+                if a == Dyn:
+                    new_symbol = Var(self.curr_symbol())
+                    new_args.append(new_symbol)
+                else:
+                    new_args.append(a)
+            return TensorType(tuple(new_args))
+        else:
+            return typ
```

Comment:
style: Can we make a general recursive map function similar to [map_aggregate](https://github.com/pytorch/pytorch/blob/c30048fccfc5914cb3924d9b2905a5e13b3e0b52/torch/fx/node.py#L537) for doing this type of stuff? That way the code for traversing type expressions and yielding a new one will all be in one place and we can update it once and all of the call-sites will have the new behavior